controllers/configurationpolicy_controller.go

```diff
@@ -1371,97 +1397,171 @@ func (r *ConfigurationPolicyReconciler) determineDesiredObjects(
 		return nil, &scopedGVR, errEvent, nil
 	}
 
+	// Fetch and filter objects using provided objectSelector
 	desiredObjects := []*unstructured.Unstructured{}
+
+	// The object either has no name specified or it is templated in the object definition
+	if desiredName == "" && objectT.ObjectSelector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(objectSelector)
+		if err != nil {
+			log.Error(err, "Failed to select the resources",
+				"objectSelector", fmt.Sprint(objectT.ObjectSelector.String()))
+
+			msg := fmt.Sprintf("Error parsing provided objectSelector: %v", err)
+
+			// only report this error if there wasn't another yet
+			errEvent := &objectTmplEvalEvent{
+				compliant: false,
+				reason:    "objectSelector error",
+				message:   msg,
+			}
+
+			return nil, &scopedGVR, errEvent, err
+		}
+
+		filteredObjects, err := r.DynamicWatcher.List(plc.ObjectIdentifier(), objGVK, "", selector)
+		if err != nil {
+			log.Error(err, "Failed to fetch the resources",
+				"objectSelector", fmt.Sprint(objectT.ObjectSelector.String()))
+
+			msg := fmt.Sprintf("Error listing resources with provided objectSelector: %v", err)
+
+			errEvent := &objectTmplEvalEvent{
+				compliant: false,
+				reason:    "objectSelector error",
+				message:   msg,
+			}
+
+			return nil, &scopedGVR, errEvent, err
+		}
+
+		// Filter objectSelector results using namespaceSelector
+		for _, res := range filteredObjects {
+			resourceNs := res.GetNamespace()
+			// Add the object if the object is in a namespace returned by the namespaceSelector
+			if _, ok := relevantNsNames[resourceNs]; ok {
+				relevantNsNames[resourceNs] = append(relevantNsNames[resourceNs], res.GetName())
+			}
+		}
+	}
+
 	hasTemplate := templates.HasTemplate(objectT.ObjectDefinition.Raw, "", true)
 	needsPerNamespaceTemplating := templateHasObjectNamespaceRegex.Match(objectT.ObjectDefinition.Raw)
 
 	var desiredObj *unstructured.Unstructured
 
-	for _, ns := range relevantNamespaces {
-		var rawDesiredObject []byte
+	for ns, names := range relevantNsNames {
+		for _, name := range names {
+			var rawDesiredObject []byte
 
-		// Process templating only on the first loop if the ObjectNamespace template variable isn't used
-		if tmplResolver != nil && hasTemplate && (desiredObj == nil || needsPerNamespaceTemplating) {
-			r.processedPolicyCache.Delete(plc.GetUID())
+			// Process templating only on the first loop if the ObjectNamespace template variable isn't used
+			if tmplResolver != nil && hasTemplate && (desiredObj == nil || needsPerNamespaceTemplating) {
+				r.processedPolicyCache.Delete(plc.GetUID())
 
-			templateContext := struct{ ObjectNamespace string }{ObjectNamespace: ns}
+				templateContext := struct{ ObjectNamespace string }{ObjectNamespace: ns}
```

Comment:
```suggestion
				templateContext := struct{ ObjectNamespace string; ObjectName string }{ObjectNamespace: ns, ObjectName: name}
```

Reply:
If you choose to not implement the `ObjectName` template variable in this PR, then the template resolution should not be per object as the resolution at the namespace level can be reused for each object.

Reply:
I think I'd prefer to tackle the ObjectName in a subsequent PR since it'd have its own tests (and there's already enough in this PR). I've set a conditional to only run on the first loop for now, and I've created an issue to track the subsequent work: https://issues.redhat.com/browse/ACM-15809